test/test.ml

```diff
@@ -0,0 +1,83 @@
+open Alcotest
+
+let bigstring_unix =
+  let empty = char_of_int 0xdf in
+  let source, drain = Unix.pipe () in
+  Unix.set_close_on_exec source;
+  Unix.set_close_on_exec drain;
+  Unix.set_nonblock source;
+  Unix.set_nonblock drain;
+  let check_read ?(shift=0) ?off ?len () =
+    let off' = match off with None -> 0 | Some off -> off in
+    let buf = Bigstring.make 50 empty in
+    let pre = Bigstring.sub buf 0 (10+off') in
+    let sub = Bigstring.sub buf 10 30 in
+    let check_buf read_len =
+      let read_sub = Bigstring.sub sub off' read_len in
+      let post = Bigstring.sub buf (10+off'+read_len) (40 - off' - read_len) in
+      Bigstring.iter (check char "pre untouched" empty) pre;
+      Bigstring.iteri (fun i c -> check char "correct read" (char_of_int (i + 97 + shift)) c) read_sub;
+      Bigstring.iter (check char "post untouched" empty) post;
+    in
+    match Bigstring_unix.read source sub ?off ?len with
+    | exception (Unix.Unix_error(Unix.EAGAIN, _, _) as e) ->
+      check_buf 0;
+      raise e
+    | exception (Unix.Unix_error(Unix.EWOULDBLOCK, _, _) as e) ->
+      check_buf 0;
+      raise e
+    | read_len ->
+      check bool "sensible read size" true (read_len <= 30 - off' && read_len >= 0);
+      check_buf read_len;
+      read_len
+  in
+  let write ?off ?len () =
+    let buf = Bytes.create 1024 in
+    while try Unix.read source buf 0 (Bytes.length buf) |> ignore; true with _ -> false do () done;
+    let count = Bigstring.init 20 (fun i -> char_of_int (i+97)) in
+    Bigstring_unix.write drain count ?off ?len
+    |> match len with None -> ignore |Some len -> check int "write length" len
+  in
+  "Bigstring_unix",
+  [ "read", `Quick, begin fun () ->
```

Comment:
calls to `Gc.compact()` can also help to find potential memory issues.

Reply:
I added a call to ``Gc.compact ()`` to ``check_buf``.